openupgrade_scripts/scripts/mass_mailing/15.0.2.5/pre-migration.py

```diff
@@ -0,0 +1,155 @@
+from openupgradelib import openupgrade
+
+
+def _rename_fields(env):
+    openupgrade.rename_columns(
+        env.cr,
+        {
+            "mailing_mailing": [
+                ("contact_ab_pc", "ab_testing_pc"),
+                ("unique_ab_testing", "ab_testing_enabled"),
+            ],
+            "mailing_trace": [
+                ("sent", "sent_datetime"),
+                ("opened", "open_datetime"),
+                ("replied", "reply_datetime"),
+                ("clicked", "links_click_datetime"),
+            ],
+        },
+    )
+
+
+def _map_mailing_mailing_reply_to_mode(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE mailing_mailing
+        SET reply_to_mode = CASE reply_to_mode
+            WHEN 'thread' THEN 'update'
+            WHEN 'email' THEN 'new'
+            ELSE reply_to_mode
+        END
+        """,
+    )
+
+
+def _map_mailing_trace_failure_type(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE mailing_trace
+        SET failure_type = CASE failure_type
+            WHEN 'SMTP' THEN 'mail_smtp'
+            WHEN 'RECIPIENT' THEN 'mail_email_invalid'
+            WHEN 'BOUNCE' THEN 'mail_email_invalid'
+            WHEN 'UNKNOWN' THEN 'unknown'
+            ELSE failure_type
+        END
+        """,
+    )
+
+
+def _map_mailing_trace_trace_status(env):
+    if not openupgrade.column_exists(env.cr, "mailing_trace", "trace_status"):
+        openupgrade.logged_query(
+            env.cr,
+            """
+                ALTER TABLE mailing_trace
+                ADD COLUMN trace_status varchar
+                """,
+        )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE mailing_trace
+        SET trace_status = CASE state
+            WHEN 'outgoing' THEN 'outgoing'
+            WHEN 'exception' THEN 'error'
+            WHEN 'sent' THEN 'sent'
+            WHEN 'opened' THEN 'open'
+            WHEN 'replied' THEN 'reply'
+            WHEN 'bounced' THEN 'bounce'
+            WHEN 'ignored' THEN 'cancel'
+            ELSE null
+        END
```

Comment:
@daonamutc1 biết đâu module nào đó thêm state mới, thì lại bị đặt thành null
```suggestion
        END
        WHERE ...
```